torch/hub.py

```diff
@@ -209,6 +212,72 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
 
     return repo_dir
 
+def get_trusted_input(repo):
+    # separate function for testing
+    response = input(
+        f"The repository {repo} does not belong to the list of trusted repositories and as such cannot be downloaded. "
+        "Do you trust this repository and wish to add it to the trusted list of repositories (y/N)?")
+    return response
+
+def _add_repo_to_trusted_list(repo, filepath, is_trusted=False, prompt=True):
+    if prompt:
+        # prompt
+        response = get_trusted_input(repo)
+    else:
+        response = "y"
+    if response.lower() in ("y", "yes"):
+        if is_trusted:
+            print("The repository is already trusted.")
+            return
+        else:
+            with open(filepath, "a") as file:
+                file.write(repo + "\n")
+
+    elif response.lower() in ("n", "no", ""):
+        raise Exception("Untrusted repository.")
+
+    else:
+        raise ValueError(f"Unrecognized response {response}.")
+
+
+def _check_repo(repo, trust_repo=None, calling_fn="load"):
+    hub_dir = get_dir()
+    filepath = os.path.join(hub_dir, "trusted_list")
+
+    if not os.path.exists(filepath):
+        Path(filepath).touch()
+
+        # Initialize with all existing repos
+        repos = next(os.walk(hub_dir))[1]
+        with open(filepath, "a") as file:
+            for _repo in repos:
+                file.write(_repo + "\n")
+
+    # load list
+    trusted_repos = list(_PREDEFINED_TRUSTED)
+    with open(filepath, 'r') as file:
+        trusted_repos += file.readlines()
+    is_trusted = any(trusted_repo == repo for trusted_repo in trusted_repos)
+
+    # to be deprecated
+    if trust_repo is None:
+        if not is_trusted:
+            warnings.warn(
+                "You are about to download an untrusted repository. In a future release, this won't be allowed. "
+                f"To add the repository to your trusted list, change the command to {calling_fn}(..., "
+                "trust_repo=False) and a command prompt will appear asking for an explicit confirmation of trust, "
+                f"or {calling_fn}(..., trust_repo=True), which will assume that the prompt is to be answered with "
+                f"'yes'.")
+        return
+
+    if not trust_repo:
+        _add_repo_to_trusted_list(repo, filepath, is_trusted)
+
+    elif trust_repo == "check" and not is_trusted:
+        _add_repo_to_trusted_list(repo, filepath)
+    elif trust_repo and not is_trusted:
+        _add_repo_to_trusted_list(repo, filepath, prompt=False)
```

Comment:
My suggestion to use a `prompt` parameter isn't very elegant after all. We have some paths that are never hit in `_add_repo_to_trusted_list()` (namely prompt=False and is_trusted=False), and it's a bit weird to call a function with a flag while this flag has just been checked before the call (is_trusted).

After all, we only want to prompt in 2 cases:

- if trust_repo=False
- if trust_repo == "check" and not is_trusted


I tried to design the code below around this observation, removing the `_add_repo_to_trusted_list()` function. WDYT? (I haven't checked whether it works or not lol) 


```py
    if (trust_repo is False) or (trust_repo == "check" and not is_trusted):
        response = input(
            f"The repository {repo} does not belong to the list of trusted repositories and as such cannot be downloaded. "
            "Do you trust this repository and wish to add it to the trusted list of repositories (y/N)?"
        )
        if response.lower() in ("y", "yes"):
            if is_trusted:
                print("The repository is already trusted.")
        elif response.lower() in ("n", "no", ""):
            raise Exception("Untrusted repository.")
        else:
            raise ValueError(f"Unrecognized response {response}.")

    # At this point we're sure that the user trusts the repo
    if not is_trusted:
        with open(filepath, "a") as file:
            file.write(repo + "\n")
```

Reply:
I agree that's way clearer than before